Adding collision filter interface to SceneGraph and Multibody Plant

SeanCurtis-TRI opened the pull request:
This is the first of two (maybe three) PRs.

Overall Plan
==========
The goal is to filter pairs of geometries for consideration in penetration queries. There are triggers that would cause geometries to not be considered: multiple geometries rigidly affixed to a single frame, disallowing collision between geometries on adjacent bodies, user-specified collision filtering. The end goal will consist of appropriate public APIs, code that automatically configures filters, and the underlying filtering mechanism.

This functionality is being spread over several PRs
1. (This PR) The public API for SceneGraph and the example of MultibodyPlant using it to prevent collisions between adjacent bodies automatically. No actual filtering takes place.
2. The core functionality that implements collision filtering. This will introduce filtered effects into the code (and unit tests will change).
3. Extend MBP's filtering API to allow user-declared collision filtering on bodies in the MultibodyPlant.

This PR in detail
=============
- `MultibodyPlant`: added filtering adjacent bodies to the `Finalize()` method.
- `SceneGraph`: public interface for specifying collision filtering between "collision groups"
- `CollisionGroup` : the definition of the collision filter declaration primitive

```
Category            added  modified  removed  
----------------------------------------------
code                455    9         0        
comments            211    2         0        
blank               133    0         0        
----------------------------------------------
TOTAL               799    11        0  
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8884)
<!-- Reviewable:end -->

ggould-tri commented:
This is a big damn PR!  Going to be a while before I finish it.

---

Review status: 16 of 17 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8884#-:-LDhOhFe9HhuQ06PO-gQ:b-cvme37)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Sorry about that; it's the first of three, I'm trying to chop one feature into somewhat tractable PRs. This has been my curse in SceneGraph -- and the curse of my reviewers, of course.

---

Review status: 16 of 17 files reviewed at latest revision, 1 unresolved discussion.

---

*[geometry/scene_graph.h, line 395 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDhJ6Pr3FQuzoTQqZ0o:-LDhSpJd8xMsF4eiuECh:b-t9ji1s) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/scene_graph.h#L395)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

`DisallowSelfCollisions()` sounds a bit more like avoiding the collision of geometries in `set` with themselves. However, as you very well explain in the intro section, self-collisions are not included in `C`. What about a name like `DisallowCollisionsBetweenSetMembers()`? or something like that? My main point being that `DissallowSelfCollisions()` seems to resonate with the idea of collisions between a geometry and itself.
</blockquote></details>

If I were to do that, I'd change `DisallowCrossCollisions` as well to something like `DisallowCollisionBetweenSets`. My various efforts all suggested to me that there was no name that really captured things nicely. So, I've decided to rely on the name + the operand + documentation.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8884#-:-LDhSeW70yEsnq5-OL4e:b2bnk3m)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 16 of 17 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[geometry/scene_graph.h, line 395 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDhJ6Pr3FQuzoTQqZ0o:-LDhqiTlBnJba3ao-_a9:bvg6er0) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/scene_graph.h#L395)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

If I were to do that, I'd change `DisallowCrossCollisions` as well to something like `DisallowCollisionBetweenSets`. My various efforts all suggested to me that there was no name that really captured things nicely. So, I've decided to rely on the name + the operand + documentation.
</blockquote></details>

I actually was going to suggest something like  `DisallowCollisionBetweenSets()` but you beat me to it. I still think `DisallowSelfCollisions()` gives the wrong idea. But I'll let you (and me) sleep on it.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8884)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Checkpointing work so far.

---

Review status: 16 of 17 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[geometry/scene_graph.h, line 395 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDhJ6Pr3FQuzoTQqZ0o:-LDl63Jz7rGKkJXKo8zt:b-xkm3ox) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/scene_graph.h#L395)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I actually was going to suggest something like  `DisallowCollisionBetweenSets()` but you beat me to it. I still think `DisallowSelfCollisions()` gives the wrong idea. But I'll let you (and me) sleep on it.
</blockquote></details>

My two cents:  If it were me, I'd go with `DisallowCollisionsWithin(things)` and `DisallowCollisionsBetween(these_things, those_things)` or something like that.  The operand is the collection, so is seems clearer to not mention the collection in the function name and let prepositions do the semantic work.

---

*[geometry/scene_graph.h, line 395 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDl7KUlAYF-3jpcqX2h:-LDl7KUlAYF-3jpcqX2i:b-bktxqy) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/scene_graph.h#L395)):*
> ```objc
>    @throws std::logic_error if the set includes ids that don't exist in the
>                             scene graph.  */
>   void DisallowSelfCollisions(const GeometrySet& set);
> ```

I think the word "Disallow" is a bit troublesome, particularly as applied to "Collisions" rather than collisions detections.  It is not clear whether the pairs are being added to C (so that collision detection can prevent penetration) or removed from C (so that they will be prevented from being considered collisions).  "Disregard" or "Ignore" would clarify that we are removing these things from C and blinding ourselves to detecting their collisions.

That said, we seem to be approaching bikeshedville on this, so I am marking this comment as "informing" and leaving it at author's discretion.  My only request is that if you keep the "disallow" language you should explicitly mention C in the method comments, since that is the language you use in the group comment.

---

*[geometry/geometry_set.h, line 47 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDl3f9S9H4hIwS4CXQW:-LDl3f9S9H4hIwS4CXQX:bvg3saz) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/geometry_set.h#L47)):*
> ```objc
>    GeometrySet geometry_set;
>    geometry_set.Add(my_ids);
>    UnaryOperation(geometry_set);
> ```

This isn't part of what the comment says ("Add a pre-existing st of ids to the group").

---

*[geometry/geometry_set.h, line 68 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDl3wAD6Kset3Zpfl1m:-LDl3wAD6Kset3Zpfl1n:bribihm) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/geometry_set.h#L68)):*
> ```objc
>    // Add a pre-existing set of ids to the group.
>    std::vector<FrameId> my_ids{...};  // Previously-defined vector of ids.
>    UnaryOperation(GeometrySet(my_ids));
> ```

As above.

---

*[geometry/geometry_set.h, line 126 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDl4jXc-T1en_ZYgTPD:-LDl4jXc-T1en_ZYgTPE:bviev43) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/geometry_set.h#L126)):*
> ```objc
>   template <typename Container>
>   explicit GeometrySet(std::initializer_list<GeometryId> geometries,
>                           const Container& frames) {
> ```

FYI indent is wrong?

---

*[geometry/geometry_set.h, line 133 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDl4vxv2CVLbYNDPwRE:-LDl4vxv2CVLbYNDPwRF:bviev43) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/geometry_set.h#L133)):*
> ```objc
>   template <typename Container>
>   explicit GeometrySet(const Container& geometries,
>                           std::initializer_list<FrameId> frames) {
> ```

FYI indent is wrong?

---

*[geometry/geometry_set.h, line 236 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDl5Zhc67QQH8SQDYgz:-LDl5Zhc67QQH8SQDYh-:b82dhjs) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/geometry_set.h#L236)):*
> ```objc
>   int num_frames() const { return static_cast<int>(frames_.size()); }
> 
>   /** returns the geometry ids in the set. */
> ```

FYI inconsistent capitalization.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8884#-:-LDl8UhtC7qJQE3-MR_J:b-cwojjd)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: 16 of 17 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8884)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>; comments below.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[geometry/test/geometry_set_test.cc, line 17 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDlAoWX2WX4-Plc8jqN:-LDlAoWX2WX4-Plc8jqO:b-qag3qe) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/test/geometry_set_test.cc#L17)):*
> ```cpp
> 
> template <typename IdContainer>
> void ExpectMembership(const GeometrySet& geometry_set, const IdContainer& ids,
> ```

BTW confusing name -- suggests it checks that the membership is exactly `ids` when in fact it just checks that `geometry_set` is a superset of `ids`.  `ExpectContains` or `ExpectContainsAll` would be clearer.

---

*[geometry/test/geometry_set_test.cc, line 73 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDlCM700R82ll2p8UOh:-LDlCM700R82ll2p8UOi:b8yykxp) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/test/geometry_set_test.cc#L73)):*
> ```cpp
>   EXPECT_EQ(set5.num_frames(), 0);
>   EXPECT_EQ(set5.num_geometries(), static_cast<int>(geometry_hash.size()));
>   ExpectMembership(set5, geometry_hash, "Unordered set constructor");
> ```

You are well past rule-of-three on this check-size-then-`ExpectMembership` pattern.  Consider factoring it.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 743 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDlDWYL49Ay8mVfHOyv:-LDlDWYL49Ay8mVfHOyw:bsl2fxj) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/multibody/multibody_tree/multibody_plant/multibody_plant.h#L743)):*
> ```objc
>   /// must be provided to the Finalize() method so that any geometric
>   /// implications of the finalization process can be appropriately handled.
>   ///
> ```

FYI it is surprising to me that scene_graph can be provided if this is not a source and will simply be ignored.  Providing a scene graph where none was involved is probably a programming error that IMO the user would wish to know about.

---

*[multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc, line 485 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDlD0Bf27nOX7RNtc07:-LDlD0Bf27nOX7RNtc08:bdx0xi8) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc#L485)):*
> ```cpp
>             "RegisterAsSourceForSceneGraph\\(\\)");
>   }
> }
> ```

FYI you didn't hit the case of not registering but finalizing a scene graph.  This case looks trivial to me, though, so author's discretion.

---

*[multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc, line 550 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDlEFDYBrAsEOEUJaHo:-LDlEFDYBrAsEOEUJaHp:b7zexr0) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc#L550)):*
> ```cpp
>   std::unique_ptr<Context<double>> context = diagram->CreateDefaultContext();
> 
>   // Set the zero configuration
> ```

FYI Complete sentence needs punctuation.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8884#-:-LDlEsC9AO11BfrDDzU8:b-7xi47h)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[geometry/scene_graph.h, line 395 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDl7KUlAYF-3jpcqX2h:-LDqDEuP7C1czbjLMdKm:blg3iae) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/scene_graph.h#L395)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Yes, I agree with @ggould-tri on mentioning C here. Note that we also forget how much context we do have in hour heads already, we must think of new users of this API, even if familiar with other engines.
I noticed yesterday that both of your methods' docs start with "Filters out.... blah.." Maybe a good name, and consistent with people's pre-conception of what filtering is, would be `FilterOutCollisionsWithin(geometry_set)` and `FilterOutCollisionsBetween(geometry_set_A, geometry_set_B)`. Maybe something to discuss with the other dyn team folks at HQ?
</blockquote></details>

(ignore this comment, I am testing a reviewable workflow feature)

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8884)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8884)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Comments mostly addressed -- still seeking feedback on the naming of the API (see comment below).

---

Review status: 9 of 17 files reviewed at latest revision, 5 unresolved discussions.

---

*[geometry/scene_graph.h, line 395 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDhJ6Pr3FQuzoTQqZ0o:-LDqV1x59C68HcFCXy0N:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/scene_graph.h#L395)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Nice! I like this one
</blockquote></details>

Done

---

*[geometry/scene_graph.h, line 395 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDl7KUlAYF-3jpcqX2h:-LDqV4p27uXpgstQUUi7:bwafce8) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/scene_graph.h#L395)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

(ignore this comment, I am testing a reviewable workflow feature)
</blockquote></details>

A counter point:

This is a slice of the final API. The final API will also include the opposite: `AllowCollisionsWithin(const GeometrySet& set)` (and, similarly,`AllowCollisionsBetween()`).  As such, I like `Disallow` and `Allow` as a natural dichotomy.  `FilterOut` doesn't seem to have as obvious opposite that flows off the tongue.

As I interpret Grant's concern, the term `disallow` is ambiguous in that it is too strong a word -- it could almost sound like some kind of constraint that effects the behavior of the simulation rather than the softer "don't report on collisions" reality. I can certainly see that and find it a sufficiently compelling issue to motivate changing the name.  How about the pair: "Ignore" and "Include" (as in `IgnoreCollisionsWithin` and `IncludeCollisionsWithin`?

Also open to other pairs of words. Once we get the verb we like, I'll update the callsites and the documentation accordingly.

---

*[geometry/test/geometry_set_test.cc, line 17 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDlAoWX2WX4-Plc8jqN:-LDqWnqM96kn5wkEF_1N:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/test/geometry_set_test.cc#L17)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

BTW confusing name -- suggests it checks that the membership is exactly `ids` when in fact it just checks that `geometry_set` is a superset of `ids`.  `ExpectContains` or `ExpectContainsAll` would be clearer.
</blockquote></details>

Done

---

*[geometry/test/geometry_set_test.cc, line 73 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDlCM700R82ll2p8UOh:-LDqb5tS20nA6MEtpRuN:b-p8xzng) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/test/geometry_set_test.cc#L73)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

You are well past rule-of-three on this check-size-then-`ExpectMembership` pattern.  Consider factoring it.
</blockquote></details>

Done (plus a bit of other scrubbing).

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 743 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDlDWYL49Ay8mVfHOyv:-LDqbPnJ-15rMkbXacCr:bn5jitu) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/multibody/multibody_tree/multibody_plant/multibody_plant.h#L743)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI it is surprising to me that scene_graph can be provided if this is not a source and will simply be ignored.  Providing a scene graph where none was involved is probably a programming error that IMO the user would wish to know about.
</blockquote></details>

This is subtle. The MBP registering as a source is indirect. As a user, I can parse an SDF file and say, "Please register the geometry in the sdf file with the scene graph", which I do by passing a non-null scenegraph into the parse function. If I do so, when I finalize, I must expect to pass in a pointer to that scene graph.

Here's the gotcha -- if the sdf doesn't have any geometry, then the MBP will never register itself as a source. But the user's code should be correct whether the SDF has geometry or not.

---

*[multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc, line 485 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDlD0Bf27nOX7RNtc07:-LDqTlYc4hPzc6zEN0yV:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc#L485)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI you didn't hit the case of not registering but finalizing a scene graph.  This case looks trivial to me, though, so author's discretion.
</blockquote></details>

Done

---

*[multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc, line 550 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDlEFDYBrAsEOEUJaHo:-LDqTplEDkNNzLYsAbIE:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc#L550)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI Complete sentence needs punctuation.
</blockquote></details>

Done

---

*[geometry/geometry_set.h, line 47 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDl3f9S9H4hIwS4CXQW:-LDqUJ5m2VcGayIpcq8s:b-cmpn5w) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/geometry_set.h#L47)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

This isn't part of what the comment says ("Add a pre-existing st of ids to the group").
</blockquote></details>

Resolved via changing the comment.

---

*[geometry/geometry_set.h, line 68 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDl3wAD6Kset3Zpfl1m:-LDqUPU6EvilDIa3MRMc:b-auma8x) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/geometry_set.h#L68)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

As above.
</blockquote></details>

Likewise as above.

---

*[geometry/geometry_set.h, line 126 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDl4jXc-T1en_ZYgTPD:-LDqUT87-aJLVmh_8zrp:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/geometry_set.h#L126)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI indent is wrong?
</blockquote></details>

Done

---

*[geometry/geometry_set.h, line 133 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDl4vxv2CVLbYNDPwRE:-LDqUUg_6WL-4RnvTEB6:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/geometry_set.h#L133)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI indent is wrong?
</blockquote></details>

Done

---

*[geometry/geometry_set.h, line 236 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDl5Zhc67QQH8SQDYgz:-LDqUaNn8bkQ65BsmFzL:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/geometry_set.h#L236)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI inconsistent capitalization.
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8884#-:-LDqnlIq7Y_ne5xSBf54:bxw55o3)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Review status: 9 of 17 files reviewed at latest revision, 2 unresolved discussions.

---

*[geometry/scene_graph.h, line 395 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDl7KUlAYF-3jpcqX2h:-LDquIcZ8Drm6z5tAMD8:bm84h9m) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/scene_graph.h#L395)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

A counter point:

This is a slice of the final API. The final API will also include the opposite: `AllowCollisionsWithin(const GeometrySet& set)` (and, similarly,`AllowCollisionsBetween()`).  As such, I like `Disallow` and `Allow` as a natural dichotomy.  `FilterOut` doesn't seem to have as obvious opposite that flows off the tongue.

As I interpret Grant's concern, the term `disallow` is ambiguous in that it is too strong a word -- it could almost sound like some kind of constraint that effects the behavior of the simulation rather than the softer "don't report on collisions" reality. I can certainly see that and find it a sufficiently compelling issue to motivate changing the name.  How about the pair: "Ignore" and "Include" (as in `IgnoreCollisionsWithin` and `IncludeCollisionsWithin`?

Also open to other pairs of words. Once we get the verb we like, I'll update the callsites and the documentation accordingly.


</blockquote></details>

How about "Include" and "Exclude"?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8884)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[geometry/scene_graph.h, line 395 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDl7KUlAYF-3jpcqX2h:-LDqx8rb0G74wG7RUz8o:bb6mt9b) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/scene_graph.h#L395)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

How about "Include" and "Exclude"?
</blockquote></details>

:+1:

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8884)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[geometry/scene_graph.h, line 395 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDl7KUlAYF-3jpcqX2h:-LDr2UT8BnkGH3q0Yq3U:b-8crohg) ([raw file](https://github.com/robotlocomotion/drake/blob/20e62daab5648e09f1674dc2df88c2b1c33d1dfe/geometry/scene_graph.h#L395)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

:+1:
</blockquote></details>

I like it too

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8884)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Changed "Disallow" to "Exclude", updated the docs. This also required some modifications to MBP to account for the `simple_gripper`. PTAL.

---

Review status: 11 of 18 files reviewed at latest revision, 1 unresolved discussion.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 235 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDg477L5mto1KmOYqTp:-LDrFWAlEdm8ZD4P9PcQ:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/3b0001c56794f7acd40c2b85798c6b60fcea9448/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L235)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

oh, in that case I believe "what" is not needed? (in "Determine what the ...")
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8884#-:-LDrErJBC2_rqNnVH_6n:b-rjmyrc)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8884)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 677 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDrfwylEypR-t8ZKdNx:-LDrfwylEypR-t8ZKdNy:bf255yi) ([raw file](https://github.com/robotlocomotion/drake/blob/5e3b30da6b1c06216a1dcb68eed01afb2a9148dd/multibody/multibody_tree/multibody_plant/multibody_plant.h#L677)):*
> ```objc
>   /// If the body with `body_index` has geometry registered with it, it returns
>   /// the geometry::FrameId associated with it. Otherwise, it returns nullopt.
>   /// @throws if called pre-finalize.
> ```

BTW any reason to disallow calling this pre-finalize? it could still be useful in that case, no?

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 678 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8884#-LDrfbWu-pex8YHeBJHc:-LDrfbWw2nBCwa7q9HGl:b468qoq) ([raw file](https://github.com/robotlocomotion/drake/blob/5e3b30da6b1c06216a1dcb68eed01afb2a9148dd/multibody/multibody_tree/multibody_plant/multibody_plant.h#L678)):*
> ```objc
>   /// the geometry::FrameId associated with it. Otherwise, it returns nullopt.
>   /// @throws if called pre-finalize.
>   optional<geometry::FrameId> GetBodyFrameIdMaybe(BodyIndex body_index) const {
> ```

we have a bunch of `MaybeGetXXX()` methods in Drake already. I suggest to stick to that convention for consistency.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8884#-:-LDrgQvD48HMs8fql5PD:bnfp4nl)*
<!-- Sent from Reviewable.io -->
